zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/util/XmlUtil.java

```diff
@@ -7,151 +7,116 @@
+      saxParserFactory.setFeature("http://xml.org/sax/features/external-general-entities", false);
+      saxParserFactory.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
+      saxParserFactory.setFeature(
+          "http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
+      return saxParserFactory;
+    } catch (final ParserConfigurationException | SAXException e) {
+      throw new RuntimeException("Error creating SAXParser: " + e.getMessage(), e);
+    }
   }
 
-  public String getFlowNodeName(final FlowNodeInstanceEntity flowNode) {
-    final var processDefinition =
-        processDefinitionServices.getByKey(flowNode.processDefinitionKey());
-    return getFlowNodeName(processDefinition, flowNode);
+  public void extractFlowNodeNames(
+      final Long processDefinitionKey, final BiConsumer<Long, ProcessFlowNode> flowNodeConsumer) {
+    final var processDefinition = processDefinitionServices.getByKey(processDefinitionKey);
+    extractFlowNodeNames(processDefinition, flowNodeConsumer);
   }
 
-  private String getFlowNodeName(
-      final ProcessDefinitionEntity processDefinition, final FlowNodeInstanceEntity flowNode) {
-    final var type = BpmnElementType.valueOf(flowNode.type().name());
-    if (type.getElementTypeName().isEmpty()) {
-      return flowNode.flowNodeId();
+  public void extractFlowNodeNames(
+      final Set<Long> processDefinitionKeys,
+      final BiConsumer<Long, ProcessFlowNode> flowNodeConsumer) {
+    final var keysList = new ArrayList<>(processDefinitionKeys);
+    final var result =
+        processDefinitionServices.search(
+            ProcessDefinitionQuery.of(q -> q.filter(f -> f.processDefinitionKeys(keysList))));
+
+    for (final var processDefinition : result.items()) {
+      extractFlowNodeNames(processDefinition, flowNodeConsumer);
     }
-    return extractFlowNodeName(
-        processDefinition.bpmnXml(),
-        processDefinition.processDefinitionId(),
-        type.getElementTypeName().get(),
-        flowNode.flowNodeId());
   }
 
-  public Map<Long, Map<String, String>> getUserTasksNames(final List<UserTaskEntity> userTasks) {
-    return getNamesForEntities(
-        userTasks,
-        UserTaskEntity::processDefinitionKey,
-        UserTaskEntity::elementId,
-        this::getUserTaskName);
+  private void extractFlowNodeNames(
+      final ProcessDefinitionEntity processDefinition,
+      final BiConsumer<Long, ProcessFlowNode> flowNodeConsumer) {
+    final var saxParserFactory = getSAXParserFactory();
+    final var is =
+        new ByteArrayInputStream(processDefinition.bpmnXml().getBytes(StandardCharsets.UTF_8));
+    final var handler = new BpmnXmlParserHandler(processDefinition, flowNodeConsumer);
+    try {
+      saxParserFactory.newSAXParser().parse(is, handler);
+    } catch (final ParserConfigurationException | SAXException | IOException e) {
+      LOG.warn("Unable to parse diagram: " + e.getMessage(), e);
+    }
   }
```

Comment:
💭 Are we doing something similar in the Camunda exporter? I believe we shouldn't implement this twice if we do. We could extract a common implementation of such an XML utility that receives a provider or function upon initialization that can physically fetch the data and a consumer that can transform the result. Those will differ for each consumer, i.e. REST API and exporter, but the underlying mechanism and caching should be the same.

Maybe this is not true but I assume there are some synergies we could leverage here.

Reply:
We already have similar stuff at multiple places (exporter, operate, tasklist and now rest).
I could check if we can extract some logic, although they all are slightly different.